src/lib/data/db-slow-query.ts

```diff
@@ -0,0 +1,106 @@
+import crypto from "crypto";
+
+class QueryTracker {
+  private key: string;
+  private sql: string;
+  private latencySum: number = 0;
+  private callCount: number = 0;
+
+  constructor(key: string, sql: string) {
+    this.key = key;
+    this.sql = sql;
+  }
+
+  public record(latency: number) {
+    this.latencySum += latency;
+    this.callCount += 1;
+  }
+
+  public getStats(): {
+    key: string;
+    sql: string;
+    count: number;
+    averageLatency: number | undefined;
+    totalLatency: number;
+  } {
+    return {
+      key: this.key,
+      sql: this.sql,
+      count: this.getCount(),
+      averageLatency: this.getAverageLatency(),
+      totalLatency: this.getTotalLatency(),
+    };
+  }
+
+  private getCount(): number {
+    return this.callCount;
+  }
+
+  private getAverageLatency(): number | undefined {
+    if (this.callCount === 0) {
+      return undefined;
+    }
+    return this.latencySum / this.callCount;
+  }
+
+  private getTotalLatency(): number {
+    return this.latencySum;
+  }
+}
+
+export class SlowQueryService {
+  private trackers: Record<string, QueryTracker>;
+  private maxLatency: number = 0;
+  private maxTotalLatency: number = 0;
+
+  constructor() {
+    this.trackers = {};
+  }
+
+  public getTs() {
+    return new Date().getTime();
+  }
+
+  public submit(sql: string, latency: number) {
+    const tracker = this.getQueryTracker(sql);
+    tracker.record(latency);
+    const { key, count, averageLatency, totalLatency } = tracker.getStats();
+    const labels = [];
+    if (latency > this.maxLatency) {
+      labels.push("SLOWEST");
+      this.maxLatency = latency;
+    }
+    if (totalLatency > this.maxTotalLatency) {
+      labels.push("MOST_IMPACT");
+      this.maxTotalLatency = totalLatency;
+    }
+    console.log(
+      JSON.stringify({
+        key,
+        latency,
+        count,
+        averageLatency,
+        totalLatency,
```

Comment:
What percentage of the total time spent in queries (across all queries) does this total latency represent?

Reply:
Change to these metrics.

- Key
- Current Latency
- Family Count
- Family Average Latency
- Family Accumulated Latency
- Total Accumulated Latency
- Share of Total
- Labels:
    - SLOWEST
        - This is when the current latency is the new max latency across all submissions.
    - MOST_IMPACT
        - This is when the family accumulated latency is the new max accumulated latency across all submissions.